### PR TITLE
Short circuit

### PR DESCRIPTION
**Pull request by tok-kkk**
**Description**

This PR includes some optimization of the order matching process.

- Check fragment status at each resolving stage. Stop resolving if it's confirmed or settled.
- Process a computation once, even if the computation is created in two epochs.
- Increase leveldb cache size and channel buffer size.
- Improve the contract address mangement.

It also fixes some tests to pass the CI and slightly improves the test coverage.

**Motivation**

Improve efficiency of the order matching process.

### CODE REVIEW

File: ome/confirmer.go [hunk 1]
```diff
@@ -172,45 +176,37 @@ func (confirmer *confirmer) checkOrdersForConfirmationFinality(orderParity order
...
+				logger.Debug(fmt.Sprintf("cannot update order fragment status, %v", err))
 			}
 		}
-
-		// Check that these orders have not already been output
-		if _, ok := confirmer.confirmed[com.Buy.OrderID]; ok {
-			continue
-		}
-		if _, ok := confirmer.confirmed[com.Sell.OrderID]; ok {
-			continue
```
Comment (loongy) [inline ome/confirmer.go]: Why have these checks been removed?

File: contract/address.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
+package contract
+
+// TokenAddresses returns the tokens for the provided network
+func TokenAddresses(network Network) map[string]string {
```
[the remark this answers is not in the document]
Comment (tok-kkk) [reply contract/address.go]: Yes, they are same with the addresses in the [contract index](https://republicprotocol.github.io/contract-index/#/)

File: leveldb/somer.go [hunk 3]
```diff
@@ -229,91 +230,138 @@ func NewSomerOrderFragmentTable(db *leveldb.DB, expiry time.Duration) *SomerOrde
 }
 
 // PutBuyOrderFragment implements the ome.OrderFragmentStorer interface.
-func (table *SomerOrderFragmentTable) PutBuyOrderFragment(epoch registry.Epoch, orderFragment order.Fragment, trader string, priority uint64) error {
+func (table *SomerOrderFragmentTable) PutBuyOrderFragment(hash [32]byte, orderFragment order.Fragment, trader string, priority uint64, status order.Status) error {
```
[the remark this answers is not in the document]
Comment (tok-kkk) [reply leveldb/somer.go]: The leveldb implementation only uses the epoch.Hash. Besides the confirmer only have the computation which just contains the epoch hash. So if we want to update the status of order fragment, we need to update the storer interface.

File: leveldb/somer.go [hunk 4]
```diff
@@ -229,91 +230,138 @@ func NewSomerOrderFragmentTable(db *leveldb.DB, expiry time.Duration) *SomerOrde
...
 // BuyOrderFragments implements the ome.OrderFragmentStorer interface.
-func (table *SomerOrderFragmentTable) BuyOrderFragments(epoch registry.Epoch) (ome.OrderFragmentIterator, error) {
-	iter := table.db.NewIterator(&util.Range{Start: table.buyKey(epoch.Hash[:], SomerBuyOrderFragmentIterBegin), Limit: table.buyKey(epoch.Hash[:], SomerBuyOrderFragmentIterEnd)}, nil)
+func (table *SomerOrderFragmentTable) BuyOrderFragments(hash [32]byte) (ome.OrderFragmentIterator, error) {
+	iter := table.db.NewIterator(&util.Range{Start: table.buyKey(hash[:], SomerBuyOrderFragmentIterBegin), Limit: table.buyKey(hash[:], SomerBuyOrderFragmentIterEnd)}, nil)
 	return newSomerOrderFragmentIterator(iter), nil
 }
 
+// UpdateBuyOrderStatus implements the ome.OrderFragmentStorer interface.
+func (table *SomerOrderFragmentTable) UpdateBuyOrderFragmentStatus(hash [32]byte, id order.ID, status order.Status) error {
```
[the remark this answers is not in the document]
Comment (tok-kkk) [reply leveldb/somer.go]: As we only have the epoch hash and orderID when we want to mark the orderfragment as confirmed. We can only get other details from the storer and update the status field, then put into the storer again .

File: ome/computation.go [hunk 5]
```diff
@@ -109,4 +109,5 @@ func (com *Computation) Equal(arg *Computation) bool {
 		com.Sell.Equal(&arg.Sell) &&
 		com.State == arg.State &&
 		com.Match == arg.Match
+	// todo : why do we want to compare state and match
```
[the remark this answers is not in the document]
Comment (tok-kkk) [reply ome/computation.go]: Only in some of our unit tests.

File: leveldb/somer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (loongy) [reply leveldb/somer.go]: Computation does have the whole fragment.

File: ome/computation.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (loongy) [reply ome/computation.go]: Ok let's leave it for now. Please format the TODO correctly. Should be `// TODO: Why do we want to compare state and match?`.

File: leveldb/somer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tok-kkk) [reply leveldb/somer.go]: We also need the trader and priority of the fragment. Added a TODO for this issue.

File: ome/computation.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tok-kkk) [reply ome/computation.go]: Done

**Pull request merged**